### PR TITLE
snap_restore: resend `recover_region` while there are TiKV restarts

### LINKED ISSUE #45206: ebs br: restore could hang if some tikv nodes are killed or restarted
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Kill some TiKV node during ebs br restore phase

### 2. What did you expect to see? (Required)

EBS BR restore continue and succeed

### 3. What did you see instead (Required)

EBS BR restore hangs

### 4. What is your TiDB version? (Required)

TiDB 6.5 and above

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #45206

Problem Summary:
When TiKV restarts, the leaders will be dropped here.

### What is changed and how it works?
We need to reregister the leaders.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fixed a bug that may cause snapshot restore stuck while there are TiKV restarts.
```

### CODE REVIEW

File: br/pkg/utils/storewatch/watching.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+func New(cli StoreMeta, cb Callback) *Watcher {
+	return &Watcher{
+		cli:        cli,
+		cb:         cb,
+		lastStores: make(map[uint64]*metapb.Store),
+	}
+}
+
+func (w *Watcher) Step(ctx context.Context) error {
+	liveStores, err := w.cli.GetAllStores(ctx, pd.WithExcludeTombstone())
```
Comment (3pointer) [inline br/pkg/utils/storewatch/watching.go:109]: It's better to use `GetAllTiKVStoresWithRetry` to skip tiflash node and add the retry mechanism

File: br/pkg/restore/data.go [hunk 2]
```diff
@@ -224,46 +263,60 @@ func (recovery *Recovery) RecoverRegions(ctx context.Context) (err error) {
...
-			return nil
+			return recovery.RecoverRegionOfStore(ectx, storeId, plan)
 		})
 	}
 	// Wait for all TiKV instances force leader and wait apply to last log.
 	return eg.Wait()
 }
 
+func (recovery *Recovery) SpawnTiKVShutDownWatchers(ctx context.Context) {
+	failedMark := map[uint64]struct{}{}
```
Comment (3pointer) [inline br/pkg/restore/data.go:278]: perfer `rebootStores`

File: br/pkg/utils/storewatch/watching.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	lastStore, ok := w.lastStores[newStore.GetId()]
+	w.lastStores[newStore.GetId()] = newStore
+	if !ok {
+		w.cb.OnNewStoreRegistered(newStore)
+		return
+	}
+	if lastStore.GetState() == metapb.StoreState_Up && newStore.GetState() == metapb.StoreState_Offline {
+		w.cb.OnDisconnect(newStore)
+	}
+	if lastStore.StartTimestamp != newStore.StartTimestamp && newStore.GetState() == metapb.StoreState_Up {
```
Comment (3pointer) [inline br/pkg/utils/storewatch/watching.go:133]: why not just use `lastStore.StartTimestamp != newStore.StartTimestamp`?

**Review by Leavrth — APPROVED** [review]

File: br/pkg/utils/storewatch/watching_test.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+			fmt.Printf("Store %d at %s Rebooted (ts = %d)\n", s.Id, s.Address, s.StartTimestamp)
+		}),
+	)
+
+	watcher := storewatch.New(cli, cb)
+	for {
+		req.NoError(watcher.Step(context.Background()))
+		time.Sleep(5 * time.Second)
+	}
+}
```
Comment (YuJuncen) [inline br/pkg/utils/storewatch/watching_test.go:138]:
```suggestion
```

**Review by 3pointer — APPROVED** [review]

File: br/pkg/restore/data.go [hunk 5]
```diff
@@ -213,6 +219,39 @@ func (recovery *Recovery) GetTotalRegions() int {
 	return len(regions)
 }
 
+func (recovery *Recovery) RecoverRegionOfStore(ctx context.Context, storeID uint64, plan []*recovpb.RecoverRegionRequest) error {
```
Comment (BornChanger) [inline br/pkg/restore/data.go:222]:
```suggestion
func (recovery *Recovery) RecoverRegionsOfStore(ctx context.Context, storeID uint64, plan []*recovpb.RecoverRegionRequest) error {
```

**Pull request merged**